### ISSUE TITLE
CustomPaint drawing stroked text with transparent body and stroke color

### ISSUE THREAD

**Issue by ArchangelXu** [flutter/flutter#112148]
## Steps to Reproduce
Tested on Android 11 and iOS 15.5.
I uploaded a simple demo, see [repo](https://github.com/ming-xi/flutter_stroke_text_issue)

Clone and run the demo, press button to change both drawBody = true and drawStroke = true. You will see that text body becomes transparent.
**Expected results:** 
CustomPaint drawing stroked text with body color and stroke color
**Actual results:** 
CustomPaint drawing stroked text with **_transparent_** body color and stroke color
<details>
<summary>Code sample</summary>

```
  void drawText(Canvas canvas, Size size) {
    if (drawStroke) {
      // draw text stroke
      p.color = Colors.black;
      p.style = PaintingStyle.stroke;
      p.strokeWidth = 1;
      drawTextInternal(canvas, size);
    }
    if (drawBody) {
      // draw text body
      p.color = textColor;
      p.style = PaintingStyle.fill;
      drawTextInternal(canvas, size);
    }
  }

  void drawTextInternal(Canvas canvas, Size size) {
    tp.text = TextSpan(
      text: "Test",
      style: TextStyle(
        fontSize: fontSize,
        // fontFamily: fontFamily,
        height: 1,
        foreground: p,
      ),
    );
    tp.layout(
      // minWidth:0,
      minWidth: 200,
      maxWidth: 200,
    );
    tp.paint(canvas, Offset(-100, -100));
  }
```

</details>

<details>
  <summary>Logs</summary>

```
```

```
Analyzing flutter_stroke_text_issue...                                  

   info • Prefer const with constant constructors • lib/main.dart:116:13 • prefer_const_constructors
   info • Prefer const with constant constructors • lib/main.dart:119:13 • prefer_const_constructors
   info • Prefer const with constant constructors • lib/main.dart:120:13 • prefer_const_constructors
   info • Prefer const with constant constructors • lib/text.dart:62:22 • prefer_const_constructors

4 issues found. (ran in 2.6s)
```

```
[✓] Flutter (Channel stable, 3.3.2, on macOS 12.0.1 21A559 darwin-x64, locale zh-Hans-CN)
    • Flutter version 3.3.2 on channel stable at /Users/xumingke/Library/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (8 days ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/xumingke/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • ANDROID_HOME = /Users/xumingke/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13F100
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Ultimate Edition (version 2021.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 70.0.3
    • Dart plugin version 213.7371

[!] Proxy Configuration
    • HTTP_PROXY is set
    • NO_PROXY is localhost,127.0.0.1
    • NO_PROXY contains localhost
    • NO_PROXY contains 127.0.0.1
    ! NO_PROXY does not contain ::1

[✓] Connected device (3 available)
    • iPhone 12 Pro (mobile) • 4604056C-8E2D-4902-AF02-7DF5CB26151F • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-5 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 12.0.1 21A559 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 105.0.5195.125

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>

<img width="324" alt="Screenshot 2022-09-22 at 6 01 46 PM" src="https://user-images.githubusercontent.com/67046386/191747798-b43c832a-87bb-453c-a05d-a6668f5c79c1.png">

**Comment by darshankawar** [comment]
Thanks for the report and code sample. Using it, I see same behavior:

<img width="317" alt="Screenshot 2022-09-22 at 5 56 45 PM" src="https://user-images.githubusercontent.com/67046386/191747941-6e9d2f12-47f3-4ec3-b4a6-6e1ebe89acb4.png">
<img width="317" alt="Screenshot 2022-09-22 at 5 57 00 PM" src="https://user-images.githubusercontent.com/67046386/191747948-2e31f702-b58c-46dc-9061-4bfdc1c4a744.png">
<img width="320" alt="Screenshot 2022-09-22 at 5 57 15 PM" src="https://user-images.githubusercontent.com/67046386/191747949-b2cb837b-801d-4406-9f96-10f842552822.png">
<img width="317" alt="Screenshot 2022-09-22 at 5 57 28 PM" src="https://user-images.githubusercontent.com/67046386/191747951-03decec9-2849-4bf9-9275-e0e286f66a63.png">


<details>
<summary> complete runnable code sample </summary>

```

import 'package:flutter/material.dart';
import 'dart:ui' as ui;


void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int style = 0;
  bool drawBody = true;
  bool drawStroke = false;

  void _changeStyle() {
    setState(() {
      style += 1;
      if (style > 2) {
        style = 0;
      }
      switch (style) {
        case 0:
          drawBody = true;
          drawStroke = false;
          break;
        case 1:
          drawBody = false;
          drawStroke = true;
          break;
        case 2:
          drawBody = true;
          drawStroke = true;

          break;
        default:
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text("drawBody   = $drawBody\ndrawStroke = $drawStroke"),
            SizedBox(
              height: 8,
            ),
            Text("press refresh button to change style"),
            SizedBox(
              height: 24,
            ),
            Container(
              width: 300,
              height: 300,
              color: Colors.red,
              child: Center(
                child: CustomPaint(
                  willChange: false,
                  painter: PostTextPainter(drawBody: drawBody, drawStroke: drawStroke),
                ),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _changeStyle,
        child: const Icon(Icons.refresh_rounded),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

class PostTextPainter extends CustomPainter {
  final Paint p = Paint();
  final Color textColor = Colors.white;
  final double fontSize = 100;
  final TextPainter tp = TextPainter(
    textDirection: TextDirection.ltr,
  );
  final bool drawBody;
  final bool drawStroke;

  PostTextPainter({this.drawBody = true, this.drawStroke = false}) {
    p.strokeCap = StrokeCap.square;
    p.strokeWidth = 1;
  }

  @override
  bool? hitTest(ui.Offset position) {
    // return super.hitTest(position);
    return false;
  }

  @override
  void paint(Canvas canvas, Size size) {
    drawText(canvas, size);
  }

  void drawText(Canvas canvas, Size size) {
    if (drawStroke) {
      // draw text stroke
      p.color = Colors.black;
      p.style = PaintingStyle.stroke;
      p.strokeWidth = 1;
      drawTextInternal(canvas, size);
    }
    if (drawBody) {
      // draw text body
      p.color = textColor;
      p.style = PaintingStyle.fill;
      drawTextInternal(canvas, size);
    }
  }

  void drawTextInternal(Canvas canvas, Size size) {
    tp.text = TextSpan(
      text: "Test",
      style: TextStyle(
        fontSize: fontSize,
        // fontFamily: fontFamily,
        height: 1,
        foreground: p,
      ),
    );
    tp.layout(
      // minWidth:0,
      minWidth: 200,
      maxWidth: 200,
    );
    tp.paint(canvas, Offset(-100, -100));
  }

  @override
  bool shouldRepaint(PostTextPainter oldDelegate) {
    return this != oldDelegate;
  }
}

```

</details>

<details>
<summary> stable, master flutter doctor -v </summary>

```

[✓] Flutter (Channel stable, 3.3.2, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.3.2 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e3c29ec00c (8 days ago), 2022-09-14 08:46:55 -0500
    • Engine revision a4ff2c53d8
    • Dart version 2.18.1
    • DevTools version 2.15.0

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[✓] Flutter (Channel master, 3.4.0-19.0.pre.438, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.4.0-19.0.pre.438 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision b2ccdb214d (81 minutes ago), 2022-09-21 23:31:23 -0400
    • Engine revision 1d5013e343
    • Dart version 2.19.0 (build 2.19.0-227.0.dev)
    • DevTools version 2.17.0
    
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by timmaffett** [comment]
Hello @ArchangelXu - I made several different pull requests to your demo repo with different possible changes that allow it to work correctly. 

- add second Paint object for the fill operation  :  https://github.com/ming-xi/flutter_stroke_text_issue/pull/1
- make a small change to the `TextStyle`'s `height` when paint is a fill :  https://github.com/ming-xi/flutter_stroke_text_issue/pull/2
- make a change to the `TextStyle`'s `overflow` when paint is a fill vs stroke :  https://github.com/ming-xi/flutter_stroke_text_issue/pull/3

It appears the the `tp.paint()` call is 'caching' the `TextSpan` painted operation and re-using it, unless it detects some change to the TextSpan (or in this case the TextStyle).   All three of the variations above make some change which is then detectable and force an TextPainter to actually paint again.

I also played with moving the paint Offset() of the second (fill) paint operation, and you can see that it does indeed paint it again, but if there is only the internal change to the Paint object `p` then it is just 'replaying' the paint operation from the first tp.paint() call.  You can see this in https://github.com/ming-xi/flutter_stroke_text_issue/pull/4

**Comment by timmaffett** [comment]
I wrote the above comment before I realized that you just can simply call `tp.markNeedsLayout()` before your call to `tp.layout(..)`.
The `TextPainter()` code is optimized to try and prevent un-needed layout operations.  The change to the paint style from stoke to fill is not 'detected' and does not force a re-layout, so the layout from the previous tp.paint() call is reused.

See https://github.com/ming-xi/flutter_stroke_text_issue/pull/5

There is a comment within `text_painter.dart`  ->
```
  // The TextPainter class should not aggressively invalidate the layout as long
  // as `markNeedsLayout` is not called (i.e., the layout cache is still valid).
  // See: https://github.com/flutter/flutter/issues/85108
```

**Comment by ArchangelXu** [comment]
> I wrote the above comment before I realized that you just can simply call `tp.markNeedsLayout()` before your call to `tp.layout(..)`. The `TextPainter()` code is optimized to try and prevent un-needed layout operations. The change to the paint style from stoke to fill is not 'detected' and does not force a re-layout, so the layout from the previous tp.paint() call is reused.
> 
> See [ming-xi/flutter_stroke_text_issue#5](https://github.com/ming-xi/flutter_stroke_text_issue/pull/5)
> 
> There is a comment within `text_painter.dart` ->
> 
> ```
>   // The TextPainter class should not aggressively invalidate the layout as long
>   // as `markNeedsLayout` is not called (i.e., the layout cache is still valid).
>   // See: https://github.com/flutter/flutter/issues/85108
> ```

Interesting... I'll try it out. Thanks~

**Comment by ArchangelXu** [comment]
@timmaffett is right, I must call TextPainter's markNeedsLayout() to make it's cache invalid:
```
TextPainter tp = TextPainter(
    textDirection: TextDirection.ltr,
);
tp.markNeedsLayout();//<---add this before layout()
tp.layout(
  minWidth: 200,
  maxWidth: 200,
);
```

**Comment by timmaffett** [comment]
The TL;DR: detail of whats going on is this:

When you call `tp.text = TextSpan(...)` with your new text span (with the same `Paint` object 'p' which has had it's style changed to to 'fill' from 'stroke' (and color...)

The setter for TextPainter.text try's to evaluate the change.  It ask the TextSpan's to compare themselves, and then in turn the TextSpan's ask their style's to compare themselves.

The style comparison code is 
```
RenderComparison compareTo(TextStyle other) {
    if (identical(this, other)) {
      return RenderComparison.identical;
    }
    if (inherit != other.inherit ||
        fontFamily != other.fontFamily ||
        fontSize != other.fontSize ||
        fontWeight != other.fontWeight ||
        fontStyle != other.fontStyle ||
        letterSpacing != other.letterSpacing ||
        wordSpacing != other.wordSpacing ||
        textBaseline != other.textBaseline ||
        height != other.height ||
        leadingDistribution != other.leadingDistribution ||
        locale != other.locale ||
        foreground != other.foreground ||
        background != other.background ||
        !listEquals(shadows, other.shadows) ||
        !listEquals(fontFeatures, other.fontFeatures) ||
        !listEquals(fontVariations, other.fontVariations) ||
        !listEquals(fontFamilyFallback, other.fontFamilyFallback) ||
        overflow != other.overflow) {
      return RenderComparison.layout;
    }
    if (color != other.color ||
        backgroundColor != other.backgroundColor ||
        decoration != other.decoration ||
        decorationColor != other.decorationColor ||
        decorationStyle != other.decorationStyle ||
        decorationThickness != other.decorationThickness) {
      return RenderComparison.paint;
    }
    return RenderComparison.identical;
  }
```

Here you can see that all that is compared is that it checks that `foreground != other.foreground`  - but you are using the *same* paint object, so this comparison is OK (because it does NOT go into the `Paint` object to look for differences).  
You need this comparison to return RenderComparison.layout if you want to 'automatically' not have to call `markNeedsLayout()` yourself.  Because your `TextSpan()` is not 'different enough' then without calling `makrNeedsLayout()` manually then the subsequent call to tp.paint() uses the previously cached info/paint operation.
